### ISSUE TITLE
Broadcasting error when using tf.keras.losses.Loss based losses instead of functions

### ISSUE THREAD

**Issue by razorx89** [tensorflow/neural-structured-learning#26]
Modified example from the introductory example (don't punish me for using a regression loss on a classification task, it is just for demonstration):
```python
import tensorflow as tf
import neural_structured_learning as nsl

# Prepare data.
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

# Create a base model -- sequential, functional, or subclass.
model = tf.keras.Sequential([
    tf.keras.Input((28, 28), name='feature'),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dense(1)
])

# Wrap the model with adversarial regularization.
adv_config = nsl.configs.make_adv_reg_config(multiplier=0.2, adv_step_size=0.05)
adv_model = nsl.keras.AdversarialRegularization(model, adv_config=adv_config)

# Compile, train with base model.
model.compile(optimizer='adam',
                  loss=tf.keras.losses.mae)
model.fit(x_train, y_train, batch_size=32, epochs=1)

model.compile(optimizer='adam',
                  loss=tf.keras.losses.MeanAbsoluteError())
model.fit(x_train, y_train, batch_size=32, epochs=1)

# Compile, train with adverserial model.
adv_model.compile(optimizer='adam',
                  loss=tf.keras.losses.mae)
adv_model.fit({'feature': x_train, 'label': y_train}, batch_size=32, epochs=1)

adv_model.compile(optimizer='adam',
                  loss=tf.keras.losses.MeanAbsoluteError())
adv_model.fit({'feature': x_train, 'label': y_train}, batch_size=32, epochs=1)
```
The first three fit invocations run fine, however, the fourth fails with a broadcasting exception:
```
Train on 60000 samples
2019-10-05 18:14:52.119047: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10.0
60000/60000 [==============================] - 2s 42us/sample - loss: 1.0129
Train on 60000 samples
60000/60000 [==============================] - 2s 38us/sample - loss: 0.6782
WARNING: Logging before flag parsing goes to stderr.
W1005 18:14:59.278777 140658904659776 training_utils.py:1348] Output output_1 missing from loss dictionary. We assume this was done on purpose. The fit and evaluate APIs will not be expecting any data to be passed to output_1.
Train on 60000 samples
60000/60000 [==============================] - 3s 55us/sample - loss: 0.7647 - mean_absolute_error: 0.6109 - adversarial_loss: 0.7689
W1005 18:15:02.706085 140658904659776 training_utils.py:1348] Output output_1 missing from loss dictionary. We assume this was done on purpose. The fit and evaluate APIs will not be expecting any data to be passed to output_1.
Train on 60000 samples
   32/60000 [..............................] - ETA: 2:31Traceback (most recent call last):
  File "/tmp/nsl_bug.py", line 36, in <module>
    adv_model.fit({'feature': x_train, 'label': y_train}, batch_size=32, epochs=1)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/training.py", line 728, in fit
    use_multiprocessing=use_multiprocessing)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/training_v2.py", line 324, in fit
    total_epochs=epochs)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/training_v2.py", line 123, in run_one_epoch
    batch_outs = execution_function(iterator)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/training_v2_utils.py", line 86, in execution_function
    distributed_function(input_fn))
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/eager/def_function.py", line 457, in __call__
    result = self._call(*args, **kwds)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/eager/def_function.py", line 503, in _call
    self._initialize(args, kwds, add_initializers_to=initializer_map)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/eager/def_function.py", line 408, in _initialize
    *args, **kwds))
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/eager/function.py", line 1848, in _get_concrete_function_internal_garbage_collected
    graph_function, _, _ = self._maybe_define_function(args, kwargs)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/eager/function.py", line 2150, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/eager/function.py", line 2041, in _create_graph_function
    capture_by_value=self._capture_by_value),
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/func_graph.py", line 915, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/eager/def_function.py", line 358, in wrapped_fn
    return weak_wrapped_fn().__wrapped__(*args, **kwds)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/training_v2_utils.py", line 73, in distributed_function
    per_replica_function, args=(model, x, y, sample_weights))
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/distribute/distribute_lib.py", line 760, in experimental_run_v2
    return self._extended.call_for_each_replica(fn, args=args, kwargs=kwargs)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/distribute/distribute_lib.py", line 1787, in call_for_each_replica
    return self._call_for_each_replica(fn, args, kwargs)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/distribute/distribute_lib.py", line 2132, in _call_for_each_replica
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/api.py", line 292, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/training_v2_utils.py", line 264, in train_on_batch
    output_loss_metrics=model._output_loss_metrics)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/training_eager.py", line 311, in train_on_batch
    output_loss_metrics=output_loss_metrics))
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/training_eager.py", line 252, in _process_single_batch
    training=training))
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/training_eager.py", line 127, in _model_loss
    outs = model(inputs, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/base_layer.py", line 847, in __call__
    outputs = call_fn(cast_inputs, *args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/autograph/impl/api.py", line 292, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/neural_structured_learning/keras/adversarial_regularization.py", line 631, in call
    gradient_tape=tape)
  File "/usr/local/lib/python3.6/dist-packages/neural_structured_learning/keras/adversarial_regularization.py", line 139, in adversarial_loss
    adv_loss = loss_fn(labels, adv_output, adv_sample_weights)
  File "/usr/local/lib/python3.6/dist-packages/neural_structured_learning/keras/adversarial_regularization.py", line 571, in _compute_total_loss
    outputs, sample_weights)
  File "/usr/local/lib/python3.6/dist-packages/neural_structured_learning/keras/adversarial_regularization.py", line 372, in _compute_loss_and_metrics
    loss_value = loss(label, output, sample_weights)
  File "/usr/local/lib/python3.6/dist-packages/neural_structured_learning/keras/adversarial_regularization.py", line 175, in __call__
    loss_value = super(_LossWrapper, self).__call__(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/losses.py", line 128, in __call__
    losses, sample_weight, reduction=self._get_reduction())
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/utils/losses_utils.py", line 107, in compute_weighted_loss
    losses, sample_weight)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/ops/losses/util.py", line 148, in scale_losses_by_sample_weight
    sample_weight = weights_broadcast_ops.broadcast_weights(sample_weight, losses)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/ops/weights_broadcast_ops.py", line 167, in broadcast_weights
    with ops.control_dependencies((assert_broadcastable(weights, values),)):
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/ops/weights_broadcast_ops.py", line 103, in assert_broadcastable
    weights_rank_static, values.shape, weights.shape))
ValueError: weights can not be broadcast to values. values.rank=0. weights.rank=2. values.shape=(). weights.shape=(32, 1).
```
This seems to happen so far only for losses subclassed from the `tf.keras.losses.Loss` base class.

**Comment by csferng** [comment]
Thanks for reporting the issue. This is the same as #21 and is fixed at b2bf5c1. The fix will be included in the next release. A temporary workaround is to set `reduction=tf.losses.Reduction.NONE` in the constructor of the `Loss` object.

**Comment by razorx89** [comment]
Great, thanks.
